### DISCUSSION TITLE
Packaging an analyzer as a NuGet package

### DISCUSSION THREAD

**Discussion by MelGrubb** [dotnet/roslyn#51662] [Q&A]
I have followed the directions outlined in the Source Generator Cookbook (https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.cookbook.md#package-a-generator-as-a-nuget-package), but I don't end up with a running analyzer. When I examine the project file of the consuming project, I have a reference to the package, but not the right kind of reference. It's just a plain old package reference, and not an analyzer reference. Is there some step missing in the documentation? Do I need to explicitly add a nuspec file and do the work in install.ps1 or something?

I'm so close to releasing my generator, but I can't until someone can simply add the NuGet package and have it start working. I really want to wrap up work on this, but have spent hours trying to figure out what I'm doing wrong here? Is there an example somewhere of creating a package out of an analyzer that I could look at?

**Comment by huoyaoyuan** [comment]
It's required to modify the csproj, setting IncludeBuildOutput to false and putting the output into analyzer path. You can find samples from CodeStyle in roslyn repo, and roslyn-analyzers repo.
I remember that the Visual Studio project template will set this for you.

**Comment by MelGrubb** [comment]
I can’t even find a project template anymore. That’s one of the first things I tried. I wanted to start over from the template and see what happens.

Is setting IncludeBuildOutput to false absolutely required? My analyzer defines an attribute that the consuming package needs to use, and I’d rather not have to build two separate packages just so that one of them can define a single attribute. I’ve gotten this to work in local tests by adding the project reference twice, once as an analyzer and once as a regular library. If setting this to false is somehow the magic that makes the installer work, then I guess that’s what I’ll have to do.

I’ll dig into the samples as well and see what I can find. Maybe this evening after work.

**Reply by huoyaoyuan** [reply]
Yes, absolutely required, but just a part. It prevents the analyzer to be resolved as a normal referece.
The project template should be available if you select Roslyn SDK workload in VS installer.

**Comment by MelGrubb** [comment]
I don’t see a Workload with that name. There is the .NET Compiler Platform SDK available individually under “Visual Studio Extension Development”. Is that it?

**Reply by jmarolf** [reply]
Thats the one

**Comment by MelGrubb** [comment]
Okay, so I can create a new project starting from the analyzer template, but the Source Generators Cookbook seems to imply that the process is far more streamlined, and that adding the item group is all it will take to package the project at build time. It sounds like what you’re telling me is that this is not the case, and I still need to do the full-blown analyzer project with its install and uninstall scripts, and all that comes with them. Is this correct?

**Reply by huoyaoyuan** [reply]
> with its install and uninstall scripts

No. Install scripts are already deprecated and can't be used in any form of nuget packages any more.

You can use [my project](https://github.com/huoyaoyuan/AntiPatternCop/blob/master/Source/AntiPatternCop.Package/AntiPatternCop.Package.csproj) as a sample.

**Reply by MelGrubb** [reply]
I've given it a look, and restructured my solution to match your example, including breaking out the package project into its own separate thing, but I'm still stuck. I have a working source generator, but the NuGet package continues to install it as a plain-old reference, and not as an analyzer.

The whole thing is up at https://github.com/MelGrubb/BuilderGenerator. There is the generator itself, which is C# only, and probably needs to be explicit about that, but I'd just like to get it working first. Then, there is the package project, similar to how you had it set up. Finally, there is the Test project, which does not directly reference the generator itself. It only knows about the generator via NuGet in order to test that part of the process. In its current state, it all works, but I had to add the ` OutputItemType="Analyzer" ReferenceOutputAssembly="false"` myself. It did not happen automatically when adding the package reference.

The BuilderGenerator.Package project builds the NuGet package, which I'm publishing to a local folder for testing.

**Comment by MelGrubb** [comment]
I’ll take a look at your project and see what I can find. The Analyzer template still creates the install and uninstall scripts, a code fix that I don’t need, a vsix package I didn’t ask for, a completely separate project for building the package, etc. It’s very heavy compared to the simple, streamlined process that the cookbook lays out. I realize that source generators are still a niche project type, but finding out what the current steps are in March 2021 has not been easy. The only references I find are out of date. Hopefully your example project will help.

**Reply by jmarolf** [reply]
I would recommend looking at https://github.com/jmarolf/generator-start for a simple getting started template for generators. I'll update it as things change but that should ge enough to create a github repo and open things in visual studio.

**Reply by MelGrubb** [reply]
I've taken a look, and gone so far as to throw out a lot of what I was doing and copy the example almost exactly, and I'm no further along than I was. Let me ask you something. Does this generator-start example produce a NuGet package that, when added to another project, installs the generator as an analyzer rather than just a simple reference? Because that's all I'm getting so far.

**Reply by jmarolf** [reply]
> Does this generator-start example produce a NuGet package that, when added to another project, installs the generator as an analyzer rather than just a simple reference? Because that's all I'm getting so far.

This is the expected behavior. In the latest release of visual studio, it will also show you the generated files. Even though it appears under the node labeled "Analyzers" everything should be working correctly.
![image](https://user-images.githubusercontent.com/9797472/111092142-78e29780-84f2-11eb-8fe0-a779c4dff426.png)

do you expect it to appear as a normal reference?

**Comment by MelGrubb** [comment]
I have a repo and a working generator. I'm just having trouble getting it into a working nuget package. I'll look at that example though and maybe it will help.Thank you

**Comment by MelGrubb** [comment]
No, but that's all I ever get. I cannot get the nuget package to install as an analyzer, only as a regular package reference. I have to tweak the csproj myself to make it an analyzer, and that change is lost if I upgrade to a newer version of the package.

**Reply by jmarolf** [reply]
> I have to tweak the csproj myself to make it an analyzer

This is correct, you will need to modify the csproj file to get it to build an analyzer nuget package. 

The requirements are:
1. use and sdk-style csproj file targeting netstandard 2.0
2. set `<IncludeBuildOutput>false</IncludeBuildOutput>` so it does not create a "normal" library reference package (like [here](https://github.com/jmarolf/generator-start/blob/c216720ea7c2bbdad4811bc79b8bfc9005686379/src/Generator/Generator.csproj#L8))
3. Create a target that packages your generator (or analyzer) according to the [analyzer package format](https://docs.microsoft.com/en-us/nuget/guides/analyzers-conventions) (like [here](https://github.com/jmarolf/generator-start/blob/main/src/Generator/Generator.csproj#L41-L45))
4. Make sure your msbuild target is called (like [here](https://github.com/jmarolf/generator-start/blob/c216720ea7c2bbdad4811bc79b8bfc9005686379/src/Generator/Generator.csproj#L26))

If you are doing all 4 things in your project then this should work without issue.

>  that change is lost if I upgrade to a newer version of the package.

I do not understand what you are saying here. Does upgrading nuget packages cause your csproj changes to become undone? That should not be happening. By what process are you upgrading nuget packages?

**Reply by sharwell** [reply]
Steps 2-4 above are now mostly obsolete. The Analyzer and Code Fix template created by the SDK creates a **\*.Package.csproj** project that handles the creation of the NuGet package.

**Comment by MelGrubb** [comment]
I believe I have tried everything you’ve suggested. I’m currently working off of Hao Yaoyuan’s "AntiPatternCop” example (https://github.com/huoyaoyuan/AntiPatternCop/blob/master/Source/AntiPatternCop.Package/AntiPatternCop.Package.csproj), so my source does not have the “TargetsForTfmSpecificContentInPackage” or “_AddGeneratorsToOutput” sections in it at the moment, but it did earlier in the day yesterday, and it made no difference. You can see my current source here https://github.com/MelGrubb/BuilderGenerator/tree/main/src. Perhaps a hybrid of the two approaches might get me somewhere.

As for the changes being lost, I am simply updating the NuGet package to a later version via Visual Studio, and it reverts to a plain PackageReference node each time. I am down to debugging the packaging itself at this point, so my “Test” project currently has no direct reference to the generator. I am publishing my package locally, and then the Test project includes it from the local NuGet folder where I published it. It is completely decoupled from the rest of the solution so that I can test the final step of the process, which is the only part that’s not working. Well, that and the fact that CodeLens is currently losing its mind… that’s a new development, and I’m hoping it goes away in “normal” use of the generator. If not, I’ll have a new thing to investigate.

**Reply by sharwell** [reply]
> and it reverts to a plain PackageReference node each time

What _exactly_ do you mean by this (before and after)?

**Reply by MelGrubb** [reply]
I mean that it used to say this

`
<PackageReference Include="BuilderGenerator" Version="0.1.10" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
`

I should point out that it only said this because I added everything after the version number by hand. Then, I bump the version number, publish a new package, update the test project, and now it says this

`
<PackageReference Include="BuilderGenerator" Version="0.1.11" />
`

At this point, every single article I see does things differently. Each person who has tried to help me has a different approach, and the template project still includes install and uninstall scripts which I'm told are obsolete now. Is there, in fact, ANY example someone can show me that actually works **today**. I just need one cohesive example that results in a usable NuGet package that installs itself as an analyzer.

**Reply by sharwell** [reply]
```
<PackageReference Include="BuilderGenerator" Version="0.1.11" PrivateAssets="all" />
```

⬆️ This is the correct form for a reference to an analyzer or source generator package. The `PrivateAssets="all"` needs to be manually added, but it should be preserved during updates.

**Reply by sharwell** [reply]
> Is there, in fact, ANY example someone can show me that actually works today. I just need one cohesive example that results in a usable NuGet package that installs itself as an analyzer.

Yes, the .NET Compiler Platform SDK has an **Analyzer and Code Fix** template which shows the current form.

**Reply by MelGrubb** [reply]
I started a new project from that template the other day. It still includes install and uninstall scripts, which I have been told are obsolete and no longer the correct way to do this. I'm just looking for a _current_ example of how to package my generator up successfully.

As for "PrivateAssets=all", I believe that is syntax that goes in the project _declaring_ the generator, not the project _consuming_ it. Also, I added it by hand, which is exactly what I'm trying to avoid in the first place, and it did _not_ survive the update to the next version. The changes were lost, and I'm back to a plain library package reference once again.

**Reply by sharwell** [reply]
> I started a new project from that template the other day. It still includes install and uninstall scripts, which I have been told are obsolete and no longer the correct way to do this. I'm just looking for a current example of how to package my generator up successfully.

The template is still the correct way. The install and uninstall scripts are not used by all project systems, but omitting them will prevent your analyzer from working properly in some scenarios. There is no downside to including them whether or not they are used.

> As for "PrivateAssets=all", I believe that is syntax that goes in the project declaring the generator, not the project consuming it.

This syntax goes in the project _using_ the analyzer or generator. There is no need to include it in the project defining the analyzer because the packaging project forcibly overrides them all with this:

```
<SuppressDependenciesWhenPacking>true</SuppressDependenciesWhenPacking>
```

> Also, I added it by hand, which is exactly what I'm trying to avoid in the first place, and it did not survive the update to the next version. The changes were lost, and I'm back to a plain library package reference once again.

This would be a bug in the NuGet Package Manager. It can be filed here:
https://github.com/NuGet/NuGet.Client

**Reply by MelGrubb** [reply]
If it's a bug in the package manager, then what I'm hearing is that you cannot, in fact, package a source generator as a NuGet package like it says you can in "Introducing Source Generators" (https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/). If that is the case, then I need to walk away and come back when it's fully baked. I have burned entire days trying to get this "last mile" to work. Everyone who has helped me has suggested something different. I have tried them all. None of them have worked. Perhaps I'll try the analyzer template, but since that means essentially starting my project over again, it's going to be a little while before I feel like doing that.

I'm walking away now. This just simply doesn't work yet.

**Reply by sharwell** [reply]
If you use the NuGet Package Manager to install an analyzer package, you should get a reference like the following (tested just now with 16.10 Preview 1):

```
<PackageReference Include="Microsoft.VisualStudio.Threading.Analyzers" Version="16.9.60">
  <PrivateAssets>all</PrivateAssets>
  <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
</PackageReference>
```

This reference should be preserved in the same form during upgrades.

Since this form is unnecessarily long, I always edit it to the form I showed above. I haven't used the NuGet Package Manager since `<PackageReference>` was introduced a few years back (I find manual edits faster, easier, and more reliable), but a user working exclusively with the NuGet Package Manager certainly shouldn't be in a situation where things "don't work". Analyzers are widely used exactly like this and reports of them not working are extremely rare.

It's probably something small we're missing. It may help to go back to the start and explain the full steps to reproduce the current behavior, along with a description of the expected behavior, the actual behavior, and an example repository we can reference.

**Reply by MelGrubb** [reply]
And that's the problem. I get a plain library package reference every single time. THAT is what I'm trying to solve. That is what doesn't work. I just want to know how to get the analyzer packaged up as a NuGet package that does what it's supposed to when installed. It's been very frustrating, and right now the idea of "start over from scratch using the template" is not what I wanted to hear. If it's the only way, then that's what I'll have to do, but it's going to be a few days before that starts to sound like fun, and I mostly suspect I'll end up back where I am right now.

What I have is a working source generator. What I want is to deliver it via NuGet. What I can't do is get it to install into the target application properly. You can see everything I'm working with here if you want. https://github.com/MelGrubb/BuilderGenerator

**Reply by sharwell** [reply]
The packaging project appears to be missing this line:
https://github.com/dotnet/roslyn-sdk/blob/7bd182146439993fe83e0310ba8f35e855126f45/src/VisualStudio.Roslyn.SDK/Roslyn.SDK/ProjectTemplates/CSharp/Diagnostic/Package/Package.csproj#L6

**Reply by sharwell** [reply]
And the BuilderGenerator.csproj project appears to be missing this line:
https://github.com/dotnet/roslyn-sdk/blob/7bd182146439993fe83e0310ba8f35e855126f45/src/VisualStudio.Roslyn.SDK/Roslyn.SDK/ProjectTemplates/CSharp/Diagnostic/Analyzer/DiagnosticAnalyzer.csproj#L5

**Reply by MelGrubb** [reply]
I have added both lines, and pushed the results up to the repo. They seem to have no net effect. I completely remove the analyzer package from the test project. Reload Visual Studio just to be sure, add the reference to the package back in, and the reference added to the test project is still a plain library reference.

**Reply by jmarolf** [reply]
@MelGrubb thanks for pointing to you project.  To test it out, I built it locally then added the resultant nuget package to a "local" feed so it would mimic being on nuget.org as closely as possible

![image](https://user-images.githubusercontent.com/9797472/111328853-daf2e800-862b-11eb-8bcd-8da728272cb8.png)

From here I was able to install the nuget package into the test project

![image](https://user-images.githubusercontent.com/9797472/111326635-f1983f80-8629-11eb-9725-d77627410bb5.png)

And everything appears to work as expected

![image](https://user-images.githubusercontent.com/9797472/111326678-fd840180-8629-11eb-943c-7d0f788e8e36.png)

`Publish-Local.ps1` is missing from the github repo so I can't verify if the `Refresh.ps1` is doing the right thing or not.  However running 

```ps1
Uninstall-Package -Id BuilderGenerator -ProjectName BuilderGenerator.Test
rm C:\Users\jmarolf\.nuget\packages\buildergenerator\$version -recurse
Install-Package -Id BuilderGenerator -ProjectName BuilderGenerator.Test
``` 

From the Nuget Manager Console in VS does work for me:

```cmd
PM> .\Refresh.ps1
Restoring packages for D:\Temp\BuilderGenerator-main\src\BuilderGenerator.Test\BuilderGenerator.Test.csproj...
Committing restore...
Writing assets file to disk. Path: D:\Temp\BuilderGenerator-main\src\BuilderGenerator.Test\obj\project.assets.json
Restored D:\Temp\BuilderGenerator-main\src\BuilderGenerator.Test\BuilderGenerator.Test.csproj (in 155 ms).
Successfully uninstalled 'BuilderGenerator 0.1.14' from BuilderGenerator.Test
Executing nuget actions took 108 ms
Time Elapsed: 00:00:00.2817399
Restoring packages for D:\Temp\BuilderGenerator-main\src\BuilderGenerator.Test\BuilderGenerator.Test.csproj...
Installed BuilderGenerator 0.1.14 from D:\Temp\BuilderGenerator-main\src\BuilderGenerator.Package\bin\Debug with content hash OXOWn1WPeiy6i5DazdUa3Kn8jnWe2oZZ387fLIR2VKmkAKVzxq14+3igBcAH6tJgJaUIHp4nlwusa6YIbW7t1Q==.
Installing NuGet package BuilderGenerator 0.1.14.
Committing restore...
Writing assets file to disk. Path:D:\Temp\BuilderGenerator-main\src\BuilderGenerator.Test\obj\project.assets.json
Restored D:\Temp\BuilderGenerator-main\src\BuilderGenerator.Test\BuilderGenerator.Test.csproj (in 322 ms).
Successfully installed 'BuilderGenerator 0.1.14' to BuilderGenerator.Test
Executing nuget actions took 153 ms
Time Elapsed: 00:00:00.5091940
```

As you've pointed out the Nuget Manager Console uninstall/install does appear to wipe out metadata. However, even with a test project file like this:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="BuilderGenerator" Version="0.1.14" />
    <PackageReference Include="NUnit" Version="3.13.1" />
    <PackageReference Include="NUnit3TestAdapter" Version="3.17.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.9.1" />
  </ItemGroup>

</Project>
```

Everything is working for me.  

What version of Visual Studio are you using and how are the results on your machine different?

**Reply by sharwell** [reply]
Adding this to a property group in the package project should resolve the failure to set `PrivateAssets`:

```
<DevelopmentDependency>true</DevelopmentDependency>
```

If so, this is unintentionally omitted from the templates in dotnet/roslyn-sdk and we should add it there after verifying it fixes the behavior of Install-Package.

**Reply by sharwell** [reply]
I submitted https://github.com/dotnet/roslyn-sdk/pull/764 to get this added to the default template.

**Reply by MelGrubb** [reply]
Adding that line finally gets the package reference to show up as something other than a plain old library reference, and the Builders are being generated as they should. There seems to be some lag in Visual Studio where making changes to the generator still requires a restart to get the new version in play. It seem like VS is caching the current version of the analyzer when it loads, and it doesn't really want to move on to the next version. I could be wrong because I can only really tell the difference by looking at the version number that I've added to the generated code attributes in the builders and I suspect that it's may just the source that's being cached and not the analyzer/generator.

Still, the Source Generators are getting better each time I look. F12 navigation to the source wasn't working until just recently, so I really appreciate the progress and all the help you've given me. I've been trying to figure out a replacement for my existing "Builder" generation because it looks like Scripty.msbuild, which I was using previously, is no longer being maintained. Before that I was using reflection-based T4 templates, but they don't seem to be able to reflect over .Net Core (now 5) code. Finally, I was using TypeWriter to generate the Builders, but that's not the greatest fit, and not really what TypeWriter was originally meant for.

I am just about ready to release this back into the wild. I had delisted my previous attempts when I discovered that the NuGet package wasn't actually working the way I thought it was. One problem remains. When looking at the "source" classes, those that the have the attribute on them, I get errors that say "Feature 'CodeLens references' is currently unavailable due to an internal error." with the following stack trace attached.

```
StreamJsonRpc.RemoteInvocationException: Object reference not set to an instance of an object.
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__139`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1.ConfiguredValueTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__17`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
RPC server exception:
System.NullReferenceException: Object reference not set to an instance of an object.
      at Microsoft.CodeAnalysis.CodeLens.CodeLensReferencesService.<TryGetMethodDescriptorAsync>d__8.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.CodeLens.CodeLensReferencesService.<>c__DisplayClass9_0.<<FindReferenceMethodsAsync>b__0>d.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.CodeLens.CodeLensReferencesService.<FindAsync>d__1`1.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.RemoteCodeLensReferencesService.<>c__DisplayClass5_0.<<FindReferenceMethodsAsync>b__0>d.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at System.Threading.Tasks.ValueTask`1.get_Result()
      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.<RunServiceImplAsync>d__12`1.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

I have verified this in a new project developed completely to the side and independent of the main solution. It is in the same repo, and is called BuilderGenerator.Demo. The only knowledge it has about the generator at all is via the NuGet package, similar to the Test project in the main solution. Opening this project and then attempting to look at the "Foo" class will eventually cause the exception and stack trace above. Looking at FooTests seems perfectly safe though. Only looking at the Foo class causes this behavior. I was seeing the same thing with the BuilderGenerator.Tests project as well.

Is this a known issue? Or is it something I'm doing wrong?

**Reply by sharwell** [reply]
> TryGetMethodDescriptorAsync

This is a known bug that has already been fixed for the next preview.

**Reply by MelGrubb** [reply]
That is excellent news. In that case, I think I'm ready to publish this thing with the caveat that the tooling itself is still in flux. I think I can call this a preview now and bump up the minor version... maybe to 0.5 since I now have reams of documentation to write.

**Comment by MelGrubb** [comment]
VS 2019, latest version. It sounds like you’re seeing what I’m seeing. I get the package reference, but it doesn’t appear to be the right kind. If you’re saying the analyzer/generator is still kicking in for you, then maybe I need to do a more isolated test. The package reference being different or more complete is the “last mile” I’ve been struggling with. If the generator is still somehow working with what ought to be incomplete syntax, then maybe things have gotten even easier when I wasn’t looking. Generators are difficult to test because I can’t tell when Visual Studio is simply holding on to the analyzer it saw first. I know that going through the Class View tab to see the generation results seems to be cached the first time you look and never ever updated after that. I have to completely restart VS in order to see a new version of the generated code, but the compiler seems to be happy with what it’s generating.

Let me give this another look tonight, and maybe I’ll discover that it’s actually been working for me even though It appeared that it shouldn’t. I should also add the PublishLocal script. It just synchronizes the various versions in the csproj files, builds and packages the project, and then publishes it to a local folder. Pretty much exactly what you’re doing. Refresh was just a testing tool to do all of that and then drop and re-add the package to the test project. As I said before, I didn’t want the test project to have any kind of reference to the generator other than the NuGet package. It seemed the cleanest way to tell whether the packaging process was working or not.

Thanks for all your help
